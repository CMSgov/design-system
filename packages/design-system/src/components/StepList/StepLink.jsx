import PropTypes from 'prop-types';
import React from 'react';

export class StepLink extends React.PureComponent {
  handleClick(event) {
    if (this.props.onClick) {
      event.preventDefault();
      this.props.onClick(this.props.href, this.props.stepId);
    }
  }

  render() {
    const { href, screenReaderText, className, children } = this.props;
    const ComponentType = this.props.component || 'a';

    return (
      <ComponentType href={href} onClick={e => this.handleClick(e)} className={className}>
        {children}
        {screenReaderText && (
          <span className="ds-u-visibility--screen-reader"> {screenReaderText}</span>
        )}
      </ComponentType>
    );
  }
}

StepLink.propTypes = {
  /**
   * Label text or HTML.
   */
  children: PropTypes.node.isRequired,
  href: PropTypes.string.isRequired,
  stepId: PropTypes.string,
  screenReaderText: PropTypes.string,
  className: PropTypes.string,
  onClick: PropTypes.func,
  component: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
};

export default StepLink;
