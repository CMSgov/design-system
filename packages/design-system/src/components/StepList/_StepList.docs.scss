/*
Step List

@responsive

A step list represents a user's progression through an application or multi-page
form. It serves as a table of contents and a way to quickly see where they are
and what they should be working on next.

Markup: steplist.example.html

Style guide: patterns.step-list
*/

/*
Start, Resume, and Edit links

A user interacts with the steps through "Start", "Resume", and "Edit"
links.

In the React component, the links' `href` properties are determined by the
`step` object's `href` property. One can also optionally pass an
`onStepLinkClick` function that will cancel the default link behavior and call
`onStepLinkClick` with `href` as a parameter for apps that handle routing with
JavaScript.

When a step has substeps and is incomplete, the `href` property should be set to
match the `href` of the current substep—that is, the first incomplete substep.
The "Edit" button will only appear on substeps that have been completed.

Style guide: patterns.step-list.buttons
*/

/*
`<StepList>`

The `StepList` component is the preferred method for building a step list, since it handles all the state logic necessary to produce its markup. A step is represented by an object with text, progress, and routing information and can optionally include an array of substeps as well as a description.

@responsive

@react-component StepList

Style guide: patterns.step-list.react
*/

/*
Step object

<table class="ds-c-table">
  <thead>
    <tr>
      <th>Name</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><code class="ds-u-font-weight--bold">completed</code></td>
      <td><code>bool</code></td>
      <td>
        <p>Whether the step has been completed</p>
      </td>
    </tr>
    <tr>
      <td><code class="ds-u-font-weight--bold">component</code></td>
      <td><code>element, func</code></td>
      <td>
        <p>When provided, this will render the passed in component for link elements in this Step. This is useful when integrating with React Router's <code>&lt;Link&gt;</code> or using your own custom component.</p>
      </td>
    </tr>
    <tr>
      <td><code class="ds-u-font-weight--bold">description</code></td>
      <td><code>string</code></td>
      <td>
        <p>Additional text to display under (only rendered for top-level steps)</p>
      </td>
    </tr>
    <tr>
      <td>
        <code class="ds-u-font-weight--bold">heading</code>
        <p><span class="ds-c-badge ds-u-bg-gray-dark">Required</span></p>
      </td>
      <td><code>string</code></td>
      <td>
        <p>Text to display as the step heading</p>
      </td>
    </tr>
    <tr>
      <td>
        <code class="ds-u-font-weight--bold">headingLevel</code>
      </td>
      <td><code>'1', '2', '3', '4', '5'</code></td>
      <td>
        <p>Heading type to override default <code>&lt;h2&gt;</code>.</p>
      </td>
    </tr>
    <tr>
      <td>
        <code class="ds-u-font-weight--bold">href</code>
        <p><span class="ds-c-badge ds-u-bg-gray-dark">Required</span></p>
      </td>
      <td><code>string</code></td>
      <td>
        <p>URL or partial URL that routes to the step. Will be passed to <code>onStepLinkClick</code> as first parameter</p>
      </td>
    </tr>
    <tr>
      <td><code class="ds-u-font-weight--bold">id</code></td>
      <td><code>string</code></td>
      <td>
        <p>Unique string representing the step. Will be passed to <code>onStepLinkClick</code> as second parameter</p>
      </td>
    </tr>
    <tr>
      <td><code class="ds-u-font-weight--bold">isNextStep</code></td>
      <td><code>bool</code></td>
      <td>
        <p>Whether this is the next unstarted step</p>
      </td>
    </tr>
    <tr>
      <td><code class="ds-u-font-weight--bold">linkText</code></td>
      <td><code>string</code></td>
      <td>
        <p>Alternative text for the link or button for this step. Will override the defaults</p>
      </td>
    </tr>
    <tr>
      <td><code class="ds-u-font-weight--bold">onClick</code></td>
      <td><code>func</code></td>
      <td>
        <p><code>onClick</code> handler for this specific step's link/button</p>
      </td>
    </tr>
    <tr>
      <td><code class="ds-u-font-weight--bold">started</code></td>
      <td><code>bool</code></td>
      <td>
        <p>Whether the step has been started</p>
      </td>
    </tr>
    <tr>
      <td><code class="ds-u-font-weight--bold">steps</code></td>
      <td><code>arrayOf[stepShape]</code></td>
      <td>
        <p>Array of substeps</p>
      </td>
    </tr>
  </tbody>
</table>

Style guide: patterns.step-list.step-object
*/

/*
---

## Managing list state

### Substeps
The `<StepList>` component takes an array of step objects. From there steps can
be broken down infinitely into sub-steps. This allows us to have unique URLs for
each part of a step; however, by default we only display two levels of this tree
—the _step_ and _substep_. This default behavior should remain unchanged except
for special circumstances. It is better not to overwhelm the user with showing
all the substeps and giving them names.

We do, nonetheless, encourage the use of sub-substeps that are not visible where
these substeps span multiple pages and have their own unique URLs. If, for
example, the user completes the first page of the `household > overall` substep
where they list the household members but has not completed the second page
where they define those members' relationships to each other, we want the
"Resume" button to take them back to the relationships page and not the first
page where they entered their names. This, of course, requires an extra
steps-building process to update a top-level steps' `href` property by
traversing the substep tree to find the first incomplete step. See the
[Completed, started, and isNextStep](#completed-started-and-isnextstep) section
below for an example JavaScript function that can change the `href` of steps
based on their substeps.

### Completed, started, and isNextStep

The _state_ of a step object will be defined for these purposes as the values
of its `completed`, `started`, and `isNextStep` properties. These correspond to
different visual states when rendered by the `<StepList>`, showing "Completed",
"Resume", or "Start" respectively. For steps with substeps, the state should be
representative of the collective states of its substeps. For example, if a step
has substeps that have `completed: false`, that step should not have
`completed: true` because not all of its substeps have been completed. Similarly
a step can only be `started` if at least one of its substeps has been `started`.
This should be true for each of the substep's substeps and so on. Below is an
example function that can propagate this state information up from the smallest
substep to the largest step before passing the steps array to the `<StepList>`
component.

```js
function propagateSubstepState(step) {
  if (step.steps) {
    const steps = step.steps.map(propagateSubstepState);
    const newStep = {
      ...step,
      steps,
      started: steps.some(s => s.started),
      completed: steps.every(s => s.completed)
    };
    if (!newStep.completed) {
      const nextStep = steps.find(s => !s.completed);
      newStep.href = nextStep.href;
    }
    return newStep;
  }
  else {
    return step;
  }
}

// ...
//
// Render function:

const steps = rawSteps.map(propagateSubstepState);
return <StepList steps={steps} />;
```

Style guide: patterns.step-list.guidance
*/
