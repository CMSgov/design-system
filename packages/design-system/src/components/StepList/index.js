export { default as StepList } from './StepList';
