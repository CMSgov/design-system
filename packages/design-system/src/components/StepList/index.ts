export * from './StepList';
export type { StepObject } from './Step';
