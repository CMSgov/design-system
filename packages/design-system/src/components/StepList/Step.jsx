import PropTypes from 'prop-types';
import React from 'react';
import StepLink from './StepLink';
import SubStep from './SubStep';
import classNames from 'classnames';

export const Step = ({ step, ...props }) => {
  if (process.env.NODE_ENV !== 'production') {
    if (step.title) {
      console.warn(
        `[Deprecated]: Please remove the 'title' prop the <StepList> step object, use 'heading' instead. This prop has been renamed and will be removed in a future release.`
      );
    }
    if (!step.title && !step.heading) {
      console.warn(
        `Please provide a 'heading' prop in the <StepList> step object, it is a required prop.`
      );
    }
  }

  const getAriaLabel = (text) => {
    const isValidTemplate = text && text.length > 0;
    const label = isValidTemplate ? text.replace('%{step}', step.heading || step.title) : undefined;
    return { 'aria-label': label };
  };
  const Heading = `h${step.headingLevel || '2'}`;
  const start = step.isNextStep;
  const resume = step.started && !step.completed;
  const className = classNames('ds-c-step', {
    'ds-c-step--current': start || resume,
    'ds-c-step--completed': step.completed,
  });
  const contentClassName = classNames('ds-c-step__content', {
    'ds-c-step__content--with-content': step.description || step.steps,
  });
  const { actionsLabelText, substepsLabelText, descriptionLabelText } = props;
  const actionsLabel = getAriaLabel(actionsLabelText);
  const substepsLabel = getAriaLabel(substepsLabelText);
  const descriptionLabel = getAriaLabel(descriptionLabelText);

  let linkLabel;
  if (step.completed && !step.steps) {
    linkLabel = step.linkText || props.editText;
  } else if (start) {
    linkLabel = step.linkText || props.startText;
  } else if (resume) {
    linkLabel = step.linkText || props.resumeText;
  }

  let linkClassName;
  if (start || resume) {
    linkClassName = 'ds-c-button ds-c-button--primary';
  }

  // TODO: make heading required after removing title
  return (
    <li className={className}>
      <div className={contentClassName}>
        <Heading className="ds-c-step__heading">{step.heading || step.title}</Heading>
        {step.description && (
          <div className="ds-c-step__description" {...descriptionLabel}>
            {step.description}
          </div>
        )}
        {step.steps && (
          <ol className="ds-c-step__substeps" {...substepsLabel}>
            {step.steps.map((s, i) => (
              <SubStep
                step={{ ...s, ...{ component: step.component || s.component } }}
                key={s.id || i}
                {...props}
              />
            ))}
          </ol>
        )}
      </div>
      <div className="ds-c-step__actions" {...actionsLabel}>
        {step.completed && <div className="ds-c-step__completed-text">{props.completedText}</div>}
        {linkLabel && (
          <StepLink
            component={step.component}
            href={step.href}
            stepId={step.id}
            screenReaderText={step.heading || step.title}
            onClick={step.onClick || props.onStepLinkClick}
            className={linkClassName}
          >
            {linkLabel}
          </StepLink>
        )}
      </div>
    </li>
  );
};

// Define the shape of a single step so we can recursively define the shape
export const stepShape = {
  id: PropTypes.string,
  href: PropTypes.string.isRequired,
  title: PropTypes.string, // [Deprecated]
  heading: PropTypes.string.isRequired,
  headingLevel: PropTypes.oneOf(['1', '2', '3', '4', '5']),
  description: PropTypes.string,
  linkText: PropTypes.string,
  completed: PropTypes.bool,
  started: PropTypes.bool,
  isNextStep: PropTypes.bool,
  onClick: PropTypes.func,
  component: PropTypes.oneOfType([PropTypes.element, PropTypes.func]),
};
stepShape.steps = PropTypes.arrayOf(PropTypes.shape(stepShape));

Step.propTypes = {
  step: PropTypes.shape(stepShape).isRequired,
  onStepLinkClick: PropTypes.func,
  showSubSubSteps: PropTypes.bool,
  completedText: PropTypes.string.isRequired,
  editText: PropTypes.string.isRequired,
  resumeText: PropTypes.string.isRequired,
  startText: PropTypes.string.isRequired,
  actionsLabelText: PropTypes.string,
  descriptionLabelText: PropTypes.string,
  substepsLabelText: PropTypes.string,
};

export default Step;
