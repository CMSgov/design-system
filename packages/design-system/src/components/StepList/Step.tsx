import React from 'react';
import StepLink, { StepLinkProps } from './StepLink';
import SubStep from './SubStep';
import classNames from 'classnames';
import { CheckIcon } from '../Icons';

type HeadingLevel = '1' | '2' | '3' | '4' | '5';

export interface StepObject {
  id?: string;
  href: string;
  title?: string; // [Deprecated]
  heading: string;
  headingLevel?: HeadingLevel;
  description?: string;
  linkText?: string;
  completed?: boolean;
  started?: boolean;
  isNextStep?: boolean;
  onClick?: StepLinkProps['onClick'];
  component?: StepLinkProps['component'];
  steps?: StepObject[];
}

export interface StepProps {
  step: StepObject;
  onStepLinkClick?: StepLinkProps['onClick'];
  showSubSubSteps?: boolean;
  completedText: string;
  editText: string;
  resumeText: string;
  startText: string;
  actionsLabelText: string;
  descriptionLabelText: string;
  substepsLabelText: string;
}

export const Step = ({ step, ...props }: StepProps) => {
  if (process.env.NODE_ENV !== 'production') {
    if (step.title) {
      console.warn(
        `[Deprecated]: Please remove the 'title' prop the <StepList> step object, use 'heading' instead. This prop has been renamed and will be removed in a future release.`
      );
    }
    if (!step.title && !step.heading) {
      console.warn(
        `Please provide a 'heading' prop in the <StepList> step object, it is a required prop.`
      );
    }
  }

  const getAriaLabel = (text) => {
    const isValidTemplate = text && text.length > 0;
    const label = isValidTemplate ? text.replace('%{step}', step.heading || step.title) : undefined;
    return { 'aria-label': label };
  };
  const Heading = `h${step.headingLevel || '2'}` as const;
  const start = step.isNextStep;
  const resume = step.started && !step.completed;
  const className = classNames('ds-c-step', {
    'ds-c-step--current': start || resume,
    'ds-c-step--completed': step.completed,
  });
  const contentClassName = classNames('ds-c-step__content', {
    'ds-c-step__content--with-content': step.description || step.steps,
  });
  const { actionsLabelText, substepsLabelText, descriptionLabelText } = props;
  const actionsLabel = getAriaLabel(actionsLabelText);
  const substepsLabel = getAriaLabel(substepsLabelText);
  const descriptionLabel = getAriaLabel(descriptionLabelText);

  let linkLabel;
  if (step.completed && !step.steps) {
    linkLabel = step.linkText || props.editText;
  } else if (start) {
    linkLabel = step.linkText || props.startText;
  } else if (resume) {
    linkLabel = step.linkText || props.resumeText;
  }

  let linkClassName;
  if (start || resume) {
    linkClassName = 'ds-c-button ds-c-button--primary';
  }

  // TODO: make heading required after removing title
  return (
    <li className={className}>
      <div className={contentClassName}>
        <Heading className="ds-c-step__heading">{step.heading || step.title}</Heading>
        {step.description && (
          <div className="ds-c-step__description" {...descriptionLabel}>
            {step.description}
          </div>
        )}
        {step.steps && (
          <ol className="ds-c-step__substeps" {...substepsLabel}>
            {step.steps.map((s, i) => (
              <SubStep
                step={{ ...s, ...{ component: step.component || s.component } }}
                key={s.id || i}
                {...props}
              />
            ))}
          </ol>
        )}
      </div>
      <div className="ds-c-step__actions" {...actionsLabel}>
        {step.completed && (
          <div className="ds-c-step__completed-text">
            <CheckIcon className="ds-c-icon-color--success" />
            {props.completedText}
          </div>
        )}
        {linkLabel && (
          <StepLink
            component={step.component}
            href={step.href}
            stepId={step.id}
            screenReaderText={step.heading || step.title}
            onClick={step.onClick || props.onStepLinkClick}
            className={linkClassName}
          >
            {linkLabel}
          </StepLink>
        )}
      </div>
    </li>
  );
};

export default Step;
