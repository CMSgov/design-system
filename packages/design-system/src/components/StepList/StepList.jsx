import PropTypes from 'prop-types';
import React from 'react';
import Step from './Step';

export const StepList = ({ steps, ...props }) => (
  <ol className="ds-c-step-list ds-u-margin-top--4">
    {steps.map((step, i) => (
      <Step
        step={{ ...step, ...{ component: props.component || step.component } }}
        key={step.id || i}
        {...props}
      />
    ))}
  </ol>
);

// Define the shape of a single step so we can recursively define the shape
// as well as reuse it in multiple components' prop-type definitions. This
// has to be in this file or else the docs generator will break.
export const stepShape = {
  id: PropTypes.string,
  href: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired, // [Deprecated]
  heading: PropTypes.string.isRequired,
  headingLevel: PropTypes.oneOf(['1', '2', '3', '4', '5']),
  description: PropTypes.string,
  linkText: PropTypes.string,
  completed: PropTypes.bool,
  started: PropTypes.bool,
  isNextStep: PropTypes.bool,
  onClick: PropTypes.func,
  component: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
};
stepShape.steps = PropTypes.arrayOf(PropTypes.shape(stepShape));

StepList.defaultProps = {
  showSubSubSteps: false,
  completedText: 'Completed',
  editText: 'Edit',
  resumeText: 'Resume',
  startText: 'Start',
  actionsLabelText: 'Primary actions for %{step}',
  descriptionLabelText: 'Description for %{step}',
  substepsLabelText: 'Secondary actions for %{step}'
};

StepList.propTypes = {
  /**
   * An array of step objects that contain information needed to render
   * them like text, state, and link/button URLs.
   * See [Start, Resume, and Edit links]({{root}}/patterns/step-list/#patterns.step-list.buttons)
   * and [Step object]({{root}}/patterns/step-list/#patterns.step-list.step-object)
   */
  steps: PropTypes.arrayOf(PropTypes.shape(stepShape)).isRequired,
  /**
   * When provided, this will render the passed in component for all link elements. This is useful when
   * integrating with React Router's `<Link>` or using your own custom component.
   * If more specific control is needed, each `step` object also accepts a `component` prop.
   */
  component: PropTypes.oneOfType([PropTypes.element, PropTypes.elementType, PropTypes.func]),
  /**
   * Whether or not to render a substep's substeps.
   */
  showSubSubSteps: PropTypes.bool,
  /**
   * Function called when a step's Edit, Start, or Resume button/link is
   * clicked. The step's `href` property will be passed as a parameter.
   */
  onStepLinkClick: PropTypes.func,
  completedText: PropTypes.string.isRequired,
  editText: PropTypes.string.isRequired,
  resumeText: PropTypes.string.isRequired,
  startText: PropTypes.string.isRequired,
  /**
   * A template string for the aria-label describing a step's actions where
   * the substring `%{step}` is replaced with that step's `heading`.
   */
  actionsLabelText: PropTypes.string.isRequired,
  /**
   * A template string for the aria-label for a step's description where
   * the substring `%{step}` is replaced with that step's `heading`.
   */
  descriptionLabelText: PropTypes.string.isRequired,
  /**
   * A template string for the aria-label describing a step's substeps where
   * the substring `%{step}` is replaced with that step's `heading`.
   */
  substepsLabelText: PropTypes.string.isRequired
};

export default StepList;
