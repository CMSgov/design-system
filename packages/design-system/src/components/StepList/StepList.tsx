import React from 'react';
import Step, { StepObject } from './Step';
import { StepLinkProps } from './StepLink';

export interface StepListProps {
  /**
   * An array of `StepObjects` that contain text, state, link URLs, and other
   * info needed to render steps.
   *
   * | Name                              | Type                      | Description                                                                                                                                                                              |
   * | --------------------------------- | ------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
   * | `completed`                       | `bool`                    | Whether the step has been completed                                                                                                                                                      |
   * | `component`                       | `element, func`           | When provided, this will render the passed in component for link elements in this Step. This is useful when integrating with React Router's `<Link>` or using your own custom component. |
   * | `description`                     | `string`                  | Additional text to dsiplay under (only rendered for top-level steps)                                                                                                                     |
   * | `heading` (required)              | `string`                  | Text to display as the step heading                                                                                                                                                      |
   * | `headingLevel`                    | `'1', '2', '3', '4', '5'` | Heading type to override default `<h2>`.                                                                                                                                                 |
   * | `href` (required)                 | `string`                  | URL or partial URL that routes to the step. Will be passed to `onStepLinkClick` as first parameter                                                                                       |
   * | `id`                              | `string`                  | Unique string representing the step. WIll be passed to `onStepLinkClick` as second parameter                                                                                             |
   * | `isNextStep`                      | `bool`                    | Whether this is the next unstarted step                                                                                                                                                  |
   * | `linkText`                        | `string`                  | Alternative text for the link or button for this step. Will override the defaults                                                                                                        |
   * | `onClick`                         | `func`                    | `onClick` handler for this specific step's link/button                                                                                                                                   |
   * | `started`                         | `bool`                    | Whether the step has been started                                                                                                                                                        |
   * | `steps`                           | `StepObject[]`            | Array of substeps                                                                                                                                                                        |
   */
  steps: StepObject[];
  /**
   * When provided, this will render the passed in component for all link elements. This is useful when
   * integrating with React Router's `<Link>` or using your own custom component.
   * If more specific control is needed, each `step` object also accepts a `component` prop.
   */
  component?: StepLinkProps['component'];
  /**
   * Whether or not to render a substep's substeps.
   */
  showSubSubSteps?: boolean;
  /**
   * Function called when a step's Edit, Start, or Resume button/link is
   * clicked. The step's `href` property will be passed as a parameter.
   */
  onStepLinkClick?: StepLinkProps['onClick'];
  completedText?: string;
  editText?: string;
  resumeText?: string;
  startText?: string;
  /**
   * A template string for the aria-label describing a step's actions where
   * the substring `%{step}` is replaced with that step's `heading`.
   */
  actionsLabelText?: string;
  /**
   * A template string for the aria-label describing a step's substeps where
   * the substring `%{step}` is replaced with that step's `heading`.
   */
  substepsLabelText?: string;
}

/**
 * For information about how and when to use this component,
 * [refer to its full documentation page](https://design.cms.gov/components/step-list/).
 *
 * ## Start, Resume, and Edit links
 *
 * A user interacts with the steps through "Start", "Resume", and "Edit"
 * links.
 *
 * In the React component, the links' `href` properties are determined by the
 * `step` object's `href` property. One can also optionally pass an
 * `onStepLinkClick` function that will cancel the default link behavior and call
 * `onStepLinkClick` with `href` as a parameter for apps that handle routing with
 * JavaScript.
 *
 * When a step has substeps and is incomplete, the `href` property should be set to
 * match the `href` of the current substep—that is, the first incomplete substep.
 * The "Edit" button will only appear on substeps that have been completed.
 *
 * ## Managing list state
 *
 * ### Substeps
 *
 * The `<StepList>` component takes an array of step objects. From there steps can
 * be broken down infinitely into sub-steps. This allows us to have unique URLs for
 * each part of a step; however, by default we only display two levels of this tree
 * —the _step_ and _substep_. This default behavior should remain unchanged except
 * for special circumstances. It is better not to overwhelm the user with showing
 * all the substeps and giving them names.
 *
 * We do, nonetheless, encourage the use of sub-substeps that are not visible where
 * these substeps span multiple pages and have their own unique URLs. If, for
 * example, the user completes the first page of the `household > overall` substep
 * where they list the household members but has not completed the second page
 * where they define those members' relationships to each other, we want the
 * "Resume" button to take them back to the relationships page and not the first
 * page where they entered their names. This, of course, requires an extra
 * steps-building process to update a top-level steps' `href` property by
 * traversing the substep tree to find the first incomplete step. See the
 * [Completed, started, and isNextStep](#completed-started-and-isnextstep) section
 * below for an example JavaScript function that can change the `href` of steps
 * based on their substeps.
 *
 * ### Completed, started, and isNextStep
 *
 * The _state_ of a step object will be defined for these purposes as the values
 * of its `completed`, `started`, and `isNextStep` properties. These correspond to
 * different visual states when rendered by the `<StepList>`, showing "Completed",
 * "Resume", or "Start" respectively. For steps with substeps, the state should be
 * representative of the collective states of its substeps. For example, if a step
 * has substeps that have `completed: false`, that step should not have
 * `completed: true` because not all of its substeps have been completed. Similarly
 * a step can only be `started` if at least one of its substeps has been `started`.
 * This should be true for each of the substep's substeps and so on. Below is an
 * example function that can propagate this state information up from the smallest
 * substep to the largest step before passing the steps array to the `<StepList>`
 * component.
 *
 * ```js
 * function propagateSubstepState(step) {
 *   if (step.steps) {
 *     const steps = step.steps.map(propagateSubstepState);
 *     const newStep = {
 *       ...step,
 *       steps,
 *       started: steps.some((s) => s.started),
 *       completed: steps.every((s) => s.completed),
 *     };
 *     if (!newStep.completed) {
 *       const nextStep = steps.find((s) => !s.completed);
 *       newStep.href = nextStep.href;
 *     }
 *     return newStep;
 *   } else {
 *     return step;
 *   }
 * }
 *
 * // ...
 * //
 * // Render function:
 *
 * const steps = rawSteps.map(propagateSubstepState);
 * return <StepList steps={steps} />;
 * ```
 */
export const StepList = ({
  steps,
  component,
  showSubSubSteps = false,
  completedText = 'Completed',
  editText = 'Edit',
  resumeText = 'Resume',
  startText = 'Start',
  ...otherProps
}: StepListProps) => (
  <ol role="list" className="ds-c-step-list">
    {steps.map((step, i) => (
      <Step
        step={{ ...step, ...{ component: component || step.component } }}
        key={step.id || i}
        {...{
          showSubSubSteps,
          completedText,
          editText,
          resumeText,
          startText,
          ...otherProps,
        }}
      />
    ))}
  </ol>
);

export default StepList;
