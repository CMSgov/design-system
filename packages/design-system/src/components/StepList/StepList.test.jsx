import React from 'react';
import StepList from './StepList';
import { generateStep } from './__mocks__/generateStep';
import { render, screen } from '@testing-library/react';

const defaultStepProps = {
  onStepLinkClick: jest.fn(),
  showSubSubSteps: false,
  completedText: 'Completed!',
  editText: 'Edit!',
  resumeText: 'Resume!',
  startText: 'Start!',
};

describe('StepList', () => {
  it('renders list of steps', () => {
    const steps = [generateStep({ id: '1' }), generateStep({ id: '2' }), generateStep({ id: 'c' })];
    const props = defaultStepProps;

    render(<StepList steps={steps} {...props} />);

    const renderedList = screen.getAllByRole('list');

    expect(renderedList.length).toEqual(1);
    expect(renderedList[0]).toHaveClass('ds-c-step-list');
    expect(renderedList[0]).toMatchSnapshot();
  });
});
