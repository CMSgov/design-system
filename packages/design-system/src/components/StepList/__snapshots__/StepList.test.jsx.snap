// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepList renders list of steps 1`] = `
<ol
  class="ds-c-step-list"
  role="list"
>
  <li
    class="ds-c-step"
    role="listitem"
  >
    <div
      class="ds-c-step__content ds-c-step__content--with-content"
    >
      <h2
        class="ds-c-step__heading"
        id="heading-1"
      >
        Do something!
      </h2>
      <div
        aria-labelledby="heading-1"
        class="ds-c-step__description"
        role="region"
      >
        Do something really cool!
      </div>
    </div>
    <div
      aria-label="Primary actions for Do something!"
      class="ds-c-step__actions"
      role="region"
    />
  </li>
  <li
    class="ds-c-step"
    role="listitem"
  >
    <div
      class="ds-c-step__content ds-c-step__content--with-content"
    >
      <h2
        class="ds-c-step__heading"
        id="heading-2"
      >
        Do something!
      </h2>
      <div
        aria-labelledby="heading-2"
        class="ds-c-step__description"
        role="region"
      >
        Do something really cool!
      </div>
    </div>
    <div
      aria-label="Primary actions for Do something!"
      class="ds-c-step__actions"
      role="region"
    />
  </li>
  <li
    class="ds-c-step"
    role="listitem"
  >
    <div
      class="ds-c-step__content ds-c-step__content--with-content"
    >
      <h2
        class="ds-c-step__heading"
        id="heading-3"
      >
        Do something!
      </h2>
      <div
        aria-labelledby="heading-3"
        class="ds-c-step__description"
        role="region"
      >
        Do something really cool!
      </div>
    </div>
    <div
      aria-label="Primary actions for Do something!"
      class="ds-c-step__actions"
      role="region"
    />
  </li>
</ol>
`;
